mitiq/rem/rem.py

```diff
@@ -0,0 +1,182 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Readout Confusion Inversion."""
+
+from typing import Callable, Optional, Union
+
+from functools import wraps
+import numpy as np
+
+from mitiq._typing import QPROGRAM, MeasurementResult
+from mitiq.executor.executor import Executor
+from mitiq.observable.observable import Observable
+
+
+def to_probability_vector(measurement: np.ndarray) -> np.ndarray:
+    """Convert a raw measurement to a probability vector.
+
+    Args:
+        measurement: A single measurement.
+    """
+    index = int("".join([str(m) for m in measurement]), base=2)
+    pv = np.zeros((2 ** measurement.shape[0]), dtype=np.uint8)
+    pv[index] = 1
+    return pv
+
+
+def sample_probability_vector(probability_vector: np.ndarray) -> np.ndarray:
+    """Sample a probability vector and return a raw measurement.
+
+    Args:
+        probability_vector: A probability vector.
+    """
+    result = np.random.choice(
+        len(probability_vector), size=1, p=probability_vector
+    )[0]
+    value = [int(i) for i in bin(result)[2:]]
+    return np.array(value, dtype=np.uint8)
+
+
+def mitigate_measurements(
+    noisy_result: MeasurementResult,
+    inverse_confusion_matrix: np.ndarray,
+) -> MeasurementResult:
+    """Applies the inverse confusion matrix against the noisy measurement
+    result and returns the adjusted measurements.
+
+    Args:
+        noisy_results: The unmitigated ``MeasurementResult``.
+        inverse_confusion_matrix: The inverse confusion matrix to apply to the
+            probability vector estimated with noisy measurement results.
+    """
+    assert noisy_result.qubit_indices is not None
+    num_qubits = len(noisy_result.qubit_indices)
+    assert inverse_confusion_matrix.shape == (2**num_qubits, 2**num_qubits)
+
+    empirical_prob_dist = np.apply_along_axis(
+        to_probability_vector, 1, noisy_result.asarray
+    )
+
+    adjusted_prob_dist = (inverse_confusion_matrix @ empirical_prob_dist.T).T
+
+    adjusted_result = np.apply_along_axis(
+        sample_probability_vector, 1, adjusted_prob_dist
+    )
+
+    result = MeasurementResult(adjusted_result, noisy_result.qubit_indices)
+    return result
+
+
+def execute_with_rem(
+    circuit: QPROGRAM,
+    executor: Union[Executor, Callable[[QPROGRAM], MeasurementResult]],
+    inverse_confusion_matrix: np.ndarray,
+    *,
+    observable: Optional[Observable] = None,
```

Comment:
For all techniques we always use` circuit, executor, observable` as the first 3 positional arguments. The other keyword-only arguments (after the `*`) are specific to the technique. 
So in this case we need something like:
```python
def execute_with_rem(
    circuit: QPROGRAM,
    executor: Union[Executor, Callable[[QPROGRAM], MeasurementResult]],
    observable: Optional[Observable] = None,    
    *,
    inverse_confusion_matrix: np.ndarray,
```

Reply:
I can make the change, but in this case we should require an inverse confusion matrix or would you rather leave it as a named parameter and throw a value error if one is not passed?

Reply:
My proposal is that the argument `inverse_confusion_matrix` should be explicitly passed as a named parameter. Since it doesn't have a default value, the user will get an error if  `inverse_confusion_matrix` is not passed.

We have the same situation with PEC, where `representations` is a necessary parameter but it's not a positional argument.

Reply:
Fixed.